sql: fix SHOW CREATE output for partitioned and interleaved partial indexes

mgartner opened the pull request:
#### sql: fix SHOW CREATE output for partitioned partial indexes

The accepted syntax for creating partitioned partial indexes requires
the `PARTITION BY` clause to precede the `WHERE` predicate clause.
Previously, `SHOW CREATE` formatted indexes with the `WHERE` clause
preceding the `PARTITION BY` clause. This commit fixes `SHOW CREATE` to
match the parser's accepted ordering of clauses.

Fixes #60019

Release note (bug fix): The `SHOW CREATE` output of a partitioned
partial index now lists the `PARTITION BY` and `WHERE` clauses in the
order accepted by the parser.

Release justification: Fixes a bug that makes `SHOW CREATE` not
round-trippable.

#### sql: fix SHOW CREATE output for partial interleaved indexes

The accepted syntax for creating partial interleaved indexes requires
the `INTERLEAVE` clause to precede the `WHERE` clause. This commit fixes
the output of partial interleaved indexes in `SHOW CREATE` to match this
syntax.

Fixes #60701

Release note (bug fix): The `SHOW CREATE` output of a partial
interleaved index now lists the `INTERLEAVED` and `WHERE` clauses in the
order accepted by the parser.

Release justification: Fixes a bug that makes `SHOW CREATE` not
round-trippable.

mgartner committed:
sql: fix SHOW CREATE output for partitioned partial indexes

The accepted syntax for creating partitioned partial indexes requires
the `PARTITION BY` clause to precede the `WHERE` predicate clause.
Previously, `SHOW CREATE` formatted indexes with the `WHERE` clause
preceding the `PARTITION BY` clause. This commit fixes `SHOW CREATE` to
match the parser's accepted ordering of clauses.

Fixes #60019

Release note (bug fix): The `SHOW CREATE` output of a partitioned
partial index now lists the `PARTITION BY` and `WHERE` clauses in the
order accepted by the parser.

Release justification: Fixes a bug that makes `SHOW CREATE` not
round-trippable.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60590)
<!-- Reviewable:end -->

otan commented on pkg/ccl/logictestccl/testdata/logic_test/partitioning_index:
```diff
@@ -184,3 +184,31 @@ inverted  CREATE TABLE public.inverted (
 -- Warning: Partitioned table with no zone configurations.
+
+# Regression test for #60019. The index predicate should be formatted after the
+# PARTITION BY clause to match the syntax that is accepted.
```
you can add a CREATE TABLE round-trip style test after https://github.com/cockroachdb/cockroach/pull/60513 merges :P oh well

otan replied:
(this is now possible!)

mgartner replied:
I added one.

otan commented on pkg/sql/crdb_internal.go:
```diff
@@ -2002,7 +2002,7 @@ func showCreateIndexWithInterleave(
 ) error {
 	f.WriteString("CREATE ")
-	idxStr, err := catformat.IndexForDisplay(ctx, table, &tableName, idx, semaCtx)
+	idxStr, err := catformat.IndexForDisplay(ctx, table, &tableName, idx, "" /* partition */, semaCtx)
```
is it possible to partition an interleave?

mgartner replied:
heh, no: https://github.com/cockroachdb/cockroach/issues/60699

but also the interleave ordering in SHOW CREATE is incorrect: https://github.com/cockroachdb/cockroach/issues/60701. I'll add another commit to this PR to fix that.

mgartner committed:
sql: fix SHOW CREATE output for partial interleaved indexes

The accepted syntax for creating partial interleaved indexes requires
the `INTERLEAVE` clause to precede the `WHERE` clause. This commit fixes
the output of partial interleaved indexes in `SHOW CREATE` to match this
syntax.

Fixes #60701

Release note (bug fix): The `SHOW CREATE` output of a partial
interleaved index now lists the `INTERLEAVED` and `WHERE` clauses in the
order accepted by the parser.

Release justification: Fixes a bug that makes `SHOW CREATE` not
round-trippable.

mgartner commented:
@otan This is ready for another look.

mgartner commented:
@otan friendly ping

otan reviewed (approved):
LGTM!

if you hit the "re-request review" button next time (the grey thing below) these float onto my github review TODO list and makes it easier to not miss!

![image](https://user-images.githubusercontent.com/3646147/108755815-d2167580-759b-11eb-8b91-2c5cfb9fed1c.png)

mgartner commented:
bors r+
